llvm/test/SYCLLowerIR/PropagateAspectUsagePass/composite_types1.ll

```diff
@@ -0,0 +1,93 @@
+; RUN: opt -passes=sycl-propagate-aspect-usage < %s -S | FileCheck %s
+;
+; Test checks simple composite structures.
+
+; Optional
+%A = type { i32 }
+
+; Not optional
+%B = type { i32 }
+
+; Not optional
+%C = type { i32 }
+
+; Contains optional - %A
+%D1 = type { %A, %B, %C }
+
+; Doesn't contain optionals
+%D2 = type { %B, %C }
+
+; Contains %D1 which contains optional %A
+%E = type { %B, %C, %D1 }
+
+; Contains a pointer on %D1 which contains optional %A
+%F1 = type { %B, %C*, %D1* }
+
+; Contains a pointer on %D2 which doesn't contain optionals
+%F2 = type { %B, %C*, %D2* }
+
+; CHECK-DAG: dso_local spir_kernel void @kernelD1() !intel_used_aspects !1 {
+define dso_local spir_kernel void @kernelD1() {
+  call spir_func void @funcD1()
+  ret void
+}
+
+; CHECK-DAG: dso_local spir_func void @funcD1() !intel_used_aspects !1 {
+define dso_local spir_func void @funcD1() {
+  %tmp = alloca %D1
+  ret void
+}
+
+; CHECK-DAG: dso_local spir_kernel void @kernelD2() {
+define dso_local spir_kernel void @kernelD2() {
+  call spir_func void @funcD2()
+  ret void
+}
+
+; CHECK-DAG: dso_local spir_func void @funcD2() {
+define dso_local spir_func void @funcD2() {
+  %tmp = alloca %D2
+  ret void
+}
+
+; CHECK-DAG: dso_local spir_kernel void @kernelE() !intel_used_aspects !1 {
+define dso_local spir_kernel void @kernelE() {
+  call spir_func void @funcE()
+  ret void
+}
+
+; CHECK-DAG: dso_local spir_func void @funcE() !intel_used_aspects !1 {
+define dso_local spir_func void @funcE() {
+  %tmp = alloca %E
+  ret void
+}
+
+; CHECK-DAG: dso_local spir_kernel void @kernelF1() !intel_used_aspects !1 {
+define dso_local spir_kernel void @kernelF1() {
+  call spir_func void @funcF1()
+  ret void
+}
+
+; CHECK-DAG: dso_local spir_func void @funcF1() !intel_used_aspects !1 {
+define dso_local spir_func void @funcF1() {
+  %tmp = alloca %F1
+  ret void
+}
+
+; CHECK-DAG: dso_local spir_kernel void @kernelF2() {
+define dso_local spir_kernel void @kernelF2() {
+  call spir_func void @funcF2()
+  ret void
+}
+
+; CHECK-DAG: dso_local spir_func void @funcF2() {
+define dso_local spir_func void @funcF2() {
+  %tmp = alloca %F2
+  ret void
+}
+
+!intel_types_that_use_aspects = !{!0}
```

Comment:
It would be great to have a handful of test cases checking that ESIMD kernels get proper aspects too.
Example SYCL ESIMD kernel and function for testing purposes:
```
define dso_local spir_kernel void @ESIMD_kernel() #0 !sycl_explicit_simd {
entry:
  ret void
}

define dso_local spir_func void @ESIMD_func() #0 !sycl_explicit_simd {
entry:
  ret void
}

attributes #0 = { "sycl-module-id"="a.cpp" }
```

Reply:
I can add such test, but I don't quite understand why we need a handful of them - aspects propagation works regardless of ESIMD presence. 

If we want to ensure that particular high-level ESIMD constructs generate correct aspects in LLVM IR - then such tests should probably be added into `sycl/` sub-project together with the patch which assigns aspects to ESIMD constructs and types

Reply:
I believe this patch should work equally well for SYCL and ESIMD kernels/functions, so not sure why ESIMD separation is needed here. At least two tests - kernel and function would be appreciated.

Reply:
Added such tests.